# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1484342392 18000
#      Fri Jan 13 16:19:52 2017 -0500
# Node ID 0c99b8c2e5bb7fe8e733f89c831274b213102389
# Parent  c2ed86feba05a5e6c2c00676f16356991c963945
additional nested function and subfunction tests

* test/bug-50014/duplicate_nested_in_subfunction_ok.m,
test/bug-50014/duplicate_nested_parent_function.m,
test/bug-50014/duplicate_parent_nested2.m,
test/bug-50014/duplicate_parent_nested_function.m,
test/bug-50014/duplicate_primary_subfunction_old_syntax.m,
test/bug-50014/duplicate_subfunction_old_syntax.m,
test/bug-50014/duplicate_subfunction_separate_scope_ok.m:
New files.

* bug-50014.tst: Update tests.

* test/bug-50014/module.mk: Update.

diff --git a/test/bug-50014/bug-50014.tst b/test/bug-50014/bug-50014.tst
--- a/test/bug-50014/bug-50014.tst
+++ b/test/bug-50014/bug-50014.tst
@@ -11,20 +11,38 @@
 ## WITHOUT ANY WARRANTY; without even the implied warranty of
 ## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 ## GNU General Public License for more details.
 ##
 ## You should have received a copy of the GNU General Public License
 ## along with Octave; see the file COPYING.  If not, see
 ## <http://www.gnu.org/licenses/>.
 
-%!error <duplicate nested function>
-%! duplicate_nested_function ();
+%!error <duplicate subfunction or nested function name>
+%! duplicate_nested_function ()
+
+%!assert (duplicate_nested_in_subfunction_ok (), 3);
 
-%!error <duplicate subfunction>
-%! duplicate_subfunction ();
+%!error <duplicate subfunction or nested function name>
+%! duplicate_nested_parent_function ()
+
+%!error <duplicate subfunction or nested function name>
+%! duplicate_parent_nested2 ()
+
+%!error <duplicate subfunction or nested function name>
+%! duplicate_parent_nested_function ()
 
-%!error <nested function duplicates primary function name>
-%! duplicate_primary_nested_function ();
+%!error <duplicate subfunction or nested function name>
+%! duplicate_primary_nested_function ()
+
+%!error <duplicate subfunction or nested function name>
+%! duplicate_primary_subfunction_old_syntax ()
 
-%!error <subfunction duplicates primary function name>
-%! duplicate_primary_subfunction ();
+%!error <duplicate subfunction or nested function name>
+%! duplicate_primary_subfunction ()
 
+%!error <duplicate subfunction or nested function name>
+%! duplicate_subfunction ()
+
+%!error <duplicate subfunction or nested function name>
+%! duplicate_subfunction_old_syntax ()
+
+%!assert (duplicate_subfunction_separate_scope_ok (), 3);
diff --git a/test/bug-50014/duplicate_nested_in_subfunction_ok.m b/test/bug-50014/duplicate_nested_in_subfunction_ok.m
new file mode 100644
--- /dev/null
+++ b/test/bug-50014/duplicate_nested_in_subfunction_ok.m
@@ -0,0 +1,19 @@
+## Test that duplicate names in separate subfunctions are OK.
+
+function retval = duplicate_nested_in_subfunction_ok ()
+  retval = sub1 () + sub2 ();
+endfunction
+
+function r = sub1 ()
+  function notbug ()
+    r = 1;
+  endfunction
+  notbug ();
+endfunction
+
+function r = sub2 ()
+  function notbug ()
+    r = 2;
+  endfunction
+  notbug ();
+endfunction
diff --git a/test/bug-50014/duplicate_nested_parent_function.m b/test/bug-50014/duplicate_nested_parent_function.m
new file mode 100644
--- /dev/null
+++ b/test/bug-50014/duplicate_nested_parent_function.m
@@ -0,0 +1,7 @@
+## Test nested function that duplicates nested function name
+function duplicate_nested_parent_function ()
+  function bug ()
+    function bug ()
+    endfunction
+  endfunction
+endfunction
diff --git a/test/bug-50014/duplicate_parent_nested2.m b/test/bug-50014/duplicate_parent_nested2.m
new file mode 100644
--- /dev/null
+++ b/test/bug-50014/duplicate_parent_nested2.m
@@ -0,0 +1,11 @@
+## Test that we detect the correct error.
+function duplicate_parent_nested2 ()
+  function notbug ()
+    function bug ()
+    endfunction
+  endfunction
+  function bug () ## no error here
+    function bug () ## error here (duplicates parent name)
+    endfunction
+  endfunction
+endfunction
diff --git a/test/bug-50014/duplicate_parent_nested_function.m b/test/bug-50014/duplicate_parent_nested_function.m
new file mode 100644
--- /dev/null
+++ b/test/bug-50014/duplicate_parent_nested_function.m
@@ -0,0 +1,7 @@
+## Test nested function that duplicates nested function name
+function duplicate_parent_nested_function ()
+  function bug ()
+    function bug ()
+    endfunction
+  endfunction
+endfunction
diff --git a/test/bug-50014/duplicate_primary_subfunction_old_syntax.m b/test/bug-50014/duplicate_primary_subfunction_old_syntax.m
new file mode 100644
--- /dev/null
+++ b/test/bug-50014/duplicate_primary_subfunction_old_syntax.m
@@ -0,0 +1,11 @@
+## Test old-style subfunctions with no END statements
+
+function r = duplicate_primary_subfunction_old_syntax ()
+  r = 0;
+  notbug ();
+
+function r = notbug ()
+  r = 1
+
+function r = duplicate_primary_subfunction_old_syntax ()
+  r = 2
diff --git a/test/bug-50014/duplicate_subfunction_old_syntax.m b/test/bug-50014/duplicate_subfunction_old_syntax.m
new file mode 100644
--- /dev/null
+++ b/test/bug-50014/duplicate_subfunction_old_syntax.m
@@ -0,0 +1,11 @@
+## Test old-style subfunctions with no END statements.
+
+function r = duplicate_subfunction_old_syntax ()
+  r = 0;
+  bug ();
+
+function r = bug ()
+  r = 1
+
+function r = bug ()
+  r = 2
diff --git a/test/bug-50014/duplicate_subfunction_separate_scope_ok.m b/test/bug-50014/duplicate_subfunction_separate_scope_ok.m
new file mode 100644
--- /dev/null
+++ b/test/bug-50014/duplicate_subfunction_separate_scope_ok.m
@@ -0,0 +1,19 @@
+## Test that nested functions with the same name are OK if they are
+## defined inside other nested functions with different names.
+
+function retval = duplicate_subfunction_separate_scope_ok ()
+  function r = suba ()
+    function sub2 ()
+      r = 1;
+    endfunction
+    sub2 ();
+  endfunction
+  function subb ()
+    function sub2 ()
+      y = 2;
+    endfunction
+    sub2 ();
+  endfunction
+  subb ();
+  retval = y + suba ();
+endfunction
diff --git a/test/bug-50014/module.mk b/test/bug-50014/module.mk
--- a/test/bug-50014/module.mk
+++ b/test/bug-50014/module.mk
@@ -1,7 +1,16 @@
 
 bug_50014_TEST_FILES = \
-  test/bug-50014/display.m \
-  test/bug-50014/bug50014.m \
-  test/bug-50014/bug-50014.tst
+  test/bug-50014/bug-50014.tst \
+  test/bug-50014/duplicate_nested_function.m \
+  test/bug-50014/duplicate_nested_in_subfunction_ok.m \
+  test/bug-50014/duplicate_nested_parent_function.m \
+  test/bug-50014/duplicate_parent_nested2.m \
+  test/bug-50014/duplicate_parent_nested_function.m \
+  test/bug-50014/duplicate_primary_nested_function.m \
+  test/bug-50014/duplicate_primary_subfunction_old_syntax.m \
+  test/bug-50014/duplicate_primary_subfunction.m \
+  test/bug-50014/duplicate_subfunction.m \
+  test/bug-50014/duplicate_subfunction_old_subfunctions.m \
+  test/bug-50014/duplicate_subfunction_separate_scope_ok.m
 
 TEST_FILES += $(bug_50014_TEST_FILES)
